cedar-policy-validator/src/custom_schema/to_json_schema.rs

```diff
@@ -0,0 +1,542 @@
+                    required: attr_decl.required.is_none(),
+                },
+            );
+        }
+        Ok(Self::Record {
+            attributes: attrs.into_iter().map(|(n, ty)| (n.node, ty)).collect(),
+            additional_attributes: false,
+        })
+    }
+}
+
+impl TryFrom<NonEmpty<Node<AppDecl>>> for ApplySpec {
+    type Error = ToJsonSchemaError;
+    fn try_from(
+        value: NonEmpty<Node<AppDecl>>,
+        lookup_func: &LookupFn,
+    ) -> Result<Self, Self::Error> {
+        // Sort out the context decls from the principal/resource decls
+        let (prs, contexts): (Vec<_>, Vec<_>) =
+            value.into_iter().partition_map(|node| match node.node {
+                AppDecl::PR(pr) => Either::Left((pr, node.loc)),
+                AppDecl::Context(context) => Either::Right((context, node.loc)),
+            });
+
+        // Sort the principal decls from the resource decls
+        let (principals, resources): (Vec<_>, Vec<_>) =
+            prs.into_iter()
+                .partition_map(|(app_decl, loc)| match app_decl.ty.node {
+                    PR::Principal => Either::Left((app_decl, loc)),
+                    PR::Resource => Either::Right((app_decl, loc)),
+                });
+
+        let principal_types = process_pr_decls(principals, "principal")?;
+
+        let resource_types = process_pr_decls(resources, "resource")?;
+
+        let attr_lookup_func = match exactly_once(contexts, "lookup_func")? {
+            Some(attrs) => Some(AttributesOrContext(SchemaType::Type(TryInto::try_into(
+                attrs,
+                lookup_func,
+            )?))),
+            None => None,
+        };
+
+        Ok(Self {
+            // In JSON schema format, unspecified resource is represented by a None field
+            resource_types: if let Some(resource_types) = resource_types {
+                Some(resource_types.into())
+            } else {
+                None
+            },
+            // In JSON schema format, unspecified principal is represented by a None field
+            principal_types: if let Some(principal_types) = principal_types {
+                Some(principal_types.into())
+            } else {
+                None
+            },
+            context: attr_lookup_func.map_or(AttributesOrContext::default(), |attrs| attrs),
+        })
+    }
+}
+
+/// If there is exactly one PRDecl here, map it to it's list of entity types
+fn process_pr_decls(
+    input: Vec<(PRAppDecl, Loc)>,
+    name: &'static str,
+) -> Result<Option<NonEmpty<SmolStr>>, ToJsonSchemaError> {
+    Ok(exactly_once(input, name)?.map(|decl| decl.entity_tys.map(|et| et.node.into())))
+}
+
+/// Fold over the list, producing the following:
+///
+/// * Err(_) -> if there is more than one element in the array
+/// * Ok(None) -> if there are no elements in the array
+/// * Ok(Some(_)) -> if there is exactly one element in the array
```

Comment:
this is `Itertools::exactly_one()` (https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.exactly_one)